daemon/sailjailclient.c

```diff
@@ -0,0 +1,921 @@
+
+    return ack;
+}
+
+/* ------------------------------------------------------------------------- *
+ * CLIENT_LAUNCH
+ * ------------------------------------------------------------------------- */
+
+static int
+client_launch_application(client_t *self)
+{
+    int               argc        = 0;
+    const gchar     **argv        = client_get_argv(self, &argc);
+    const char       *binary      = path_basename(*argv);
+    const gchar      *desktop     = client_get_desktop_path(self);
+    gchar            *application = path_to_desktop_name(desktop);
+
+    /* Prompt for launch permission */
+    if( !client_prompt_permissions(self, application) )
+        goto EXIT;
+
+    const gchar **granted = client_get_granted(self);
+
+    /* Check if privileged launch is needed / possible */
+    bool privileged = false;
+    for( int i = 0; !privileged && granted[i]; ++i )
+        privileged = !strcmp(granted[i], "Privileged");
+
+    if( privileged && !client_is_privileged(self) ) {
+        log_err("privileged launch is needed but not possible");
+        goto EXIT;
+    }
+
+    /* Obtain application details */
+    if( !client_query_appinfo(self, application) )
+        goto EXIT;
+
+    const char  *exec        = client_deskop_exec(self);
+    const char  *org_name    = client_sailjail_organization_name(self);
+    const char  *app_name    = client_sailjail_application_name(self);
+    const char **permissions = client_sailjail_application_permissions(self);
+    const char  *service     = client_maemo_service(self);
+    const char  *method      = client_maemo_method(self);
+
+    if( log_p(LOG_DEBUG) ) {
+        log_debug("exec     = %s", exec);
+        log_debug("org_name = %s", org_name);
+        log_debug("app_name = %s", app_name);
+        log_debug("service  = %s", service);
+        log_debug("method   = %s", method);
+        for( int i = 0; permissions && permissions[i]; ++i )
+            log_debug("permissions += %s", permissions[i]);
+        for( int i = 0; granted && granted[i]; ++i )
+            log_debug("granted     += %s", granted[i]);
+    }
+
+    /* Check that command line we have matches Exec line in desktop file */
+    if( !exec ) {
+        log_err("Exec line not defined");
+        goto EXIT;
+    }
+
+    if( !sailjailclient_validate_argv(argc, argv, exec) ) {
+        log_err("Command line does not match template");
+        goto EXIT;
+    }
+
+    /* Construct firejail command line to execute */
+    client_add_firejail_option(self, "/usr/bin/firejail");
+    client_add_firejail_option(self, "--quiet");
+    client_add_firejail_option(self, "--private-bin=%s", binary);
+    client_add_firejail_option(self, "--whitelist=/usr/share/%s", binary);
+    client_add_firejail_option(self, "--whitelist=%s", desktop);
+
+    /* Legacy app binary based data directories are made available.
+     * But only if they already exist. */
+    client_add_firejail_directory(self, false, "${HOME}/.local/share/%s", binary);
```

Comment:
This is a new feature. Probably much needed one too.

Reply:
I've not tracked where it comes from, but it does show up in constructed argv i.e. it should not be a new feature.

Reply:
Oh. Well in either case that's good for migrating data and we need that for legacy app sandboxing support. Though I believe we need to extend it to some other paths (to .config and .cache) as well then. And have the actual directories (--mkdir) iff the legacy app sandboxing is active. However, that's not part of this task.